### DISCUSSION TITLE
How "Download folder" is implemented?

### DISCUSSION THREAD

**Discussion by theaccessvirus** [nicotine-plus/nicotine-plus#2583] [General]
Unfortunately i'm not well versed with python so i'm trying to implement a small application myself using [this node.js client](https://github.com/f-hj/slsk-client) which still don't have the "download folder" feature implemented.

I'm wondering if someone could kindly point me to the python file and/or protocol and/or logic for such implementation in a client?

I understand once i find a file in a search result i have the username, the file path ( and other metadata ) but i don't understand how to request a list of files from such folder and then download them sequentially.

any help is more than appreciated,

thank you

**Comment by slook** [comment]
See:
- https://github.com/nicotine-plus/nicotine-plus/blob/master/doc/SLSKPROTOCOL.md#foldercontentsrequest
- https://github.com/nicotine-plus/nicotine-plus/blob/master/doc/SLSKPROTOCOL.md#foldercontentsresponse

Current implementions:
- _Browse Shares:_
https://github.com/nicotine-plus/nicotine-plus/blob/70167b9d3964688486d2bc13a8c6eefcb2a39a9f/pynicotine/userbrowse.py#L210-L241

- _Search Files:_
https://github.com/nicotine-plus/nicotine-plus/blob/70167b9d3964688486d2bc13a8c6eefcb2a39a9f/pynicotine/gtkgui/search.py#L1418-L1456
https://github.com/nicotine-plus/nicotine-plus/blob/70167b9d3964688486d2bc13a8c6eefcb2a39a9f/pynicotine/search.py#L82-L92
https://github.com/nicotine-plus/nicotine-plus/blob/70167b9d3964688486d2bc13a8c6eefcb2a39a9f/pynicotine/transfers.py#L1635-L1636

Note if PR https://github.com/nicotine-plus/nicotine-plus/pull/2538 (or something similar) is implemented, then the methods in the search implementation are planned to change to something better.

N+ requires improvement in this area, not enough flexibility is offered and it is not performant enough on a large selection (contributions are welcome). Good luck with your project!
